plugins/outputs/azure_data_explorer/azure_data_explorer.go

```diff
@@ -75,22 +68,35 @@ func (adx *AzureDataExplorer) Connect() error {
 		Authorizer: authorizer,
 	}
 	client, err := kusto.New(adx.Endpoint, authorization)
-
+	adx.Log.Debug("Connect : Client initialized successfully")
 	if err != nil {
 		return err
 	}
 	adx.client = client
-	adx.ingesters = make(map[string]localIngestor)
-	adx.createIngestor = createRealIngestor
-
+	adx.ingestors = make(map[string]ingest.Ingestor)
 	return nil
 }
 
+// Clean up and close the ingestor
 func (adx *AzureDataExplorer) Close() error {
+	var err error
+	for _, v := range adx.ingestors {
+		err = v.Close()
+	}
+	err2 := adx.client.Close()
+	if err == nil {
+		err = err2
+	} else {
+		err = kustoerrors.GetCombinedError(err, err2)
+	}
+	if err != nil {
+		adx.Log.Warn("error closing connections")
+	} else {
+		adx.Log.Info("closed ingestor and client")
+	}
```

Comment:
This is hard to read, can you maybe write down the combinatorics here, comment each branch and do the right thing (e.g. return the error early)?